cli/src/command_error.rs

```diff
@@ -877,8 +877,8 @@ fn revset_resolution_error_hint(err: &RevsetResolutionError) -> Option<String> {
 fn string_pattern_parse_error_hint(err: &StringPatternParseError) -> Option<String> {
     match err {
         StringPatternParseError::InvalidKind(_) => Some(
-            "Try prefixing with one of `exact:`, `glob:`, `regex:`, `substring:`, or one of these \
-             with `-i` added (e.g. `glob-i:`) for case-insensitive matching"
+            "Try prefixing with one of `exact:`, `glob:`, `regex:`, `sub:`, or one of these with \
+             `-i` added (e.g. `glob-i:`) for case-insensitive matching"
```

Comment:
I don't know if `sub:` is common short name for substring. Even if it is, I wouldn't change the error message to use the short form.

Reply:
I guess `substr` and `substr-i` is another option.

Yet another option would be to allow single-letter versions for all of them: `e:`, `g:`, `e-i:`, `r:`, etc.

My thinking was to eventually softly (or not-so-softly) deprecate the long `substring:` with `sub:` (or `substr:`). But I don't feel strongly about it.

For something like `description()`, shortening `substring-i:` to `i:` would make sense. However, as string patterns default to `exact:` in bookmark names, it would be confusing in that case.

Reply:
`substr` sounds better than `sub`. It wouldn't save much typing, though.

iirc, there was an idea to making `glob:` the default in both CLI and revsets, Then, `i:` could be aliased to `glob-i:`?

Reply:
Yes, I still like the `glob:`/`i:` plan. We could maybe even do `regex:` these days but I think `glob:` would be less invasive.

Reply:
I think `substr` is better than `substring`, so I'll do that. Let me know if you prefer that I get rid of that commit entirely (or move it to a separate PR).

For the `glob:` and "i:" plan, remind me, does `jj bookmark move feat` match all strings containing `feat` or only the string "feat" exactly? (We should document it either way, I could add a commit). In the latter case, I think I like it; it's what I've been assuming naturally before reading the docs. Though, of course, some users who like substring matches will be inconvenienced.

However, if that happens, we have the option of adding an `s:` alias for `substring:`.

Reply:
> I think `substr` is better than `substring`, so I'll do that. Let me know if you prefer that I get rid of that commit entirely (or move it to a separate PR).

I don't mind if you think `substr` is strictly better. My feeling is that `substr` isn't so different from `substring`, and I personally would avoid paying extra (documenting) cost for that.

> For the `glob:` and `i:` plan, remind me, does `jj bookmark move feat` match all strings containing `feat` or only the string "feat" exactly?

The latter. There will be a bigger behavior change in revset. You'll need to write `subject('*foo*')` instead of `subject(foo)`. (The parser might have to accept unquoted `*foo*`.)

Reply:
I prefer `substring` as the canonical form but I don't mind adding `substr` as a hidden alias for `substring` to help people who are used to that form from standard library (e.g. C++ or JavaScript). Actually, if we wanted to match standard libraries, it seems like `contains` is a better match (I think `substr()` usually a range of indexes as input).

Reply:
I like `contains:`. I think I proposed that around when we were talking about defaults. I forget what the reception was.

Reply:
Yeah, I’d personally like to try `glob:` as default across the board and `i:` as short for `glob-i:`, and then see if we still feel the need for shorter names. I intended to push for this last year but didn’t get around to it. IIRC there was some concern about being able to refer to bookmarks with `*` in them or the like, but I think that’s probably not a big deal.

Potential issues with default `glob:` I can foresee:

* Searching for commit messages with a `?` in them. We could/should probably turn off `?` if it’s an option, since it’s pretty niche.
* Searching for commit messages with `[…]` in them, e.g. `[topic] …` instead of `topic: …`. Not sure if there’s anything sensible we could do about that.